### PR TITLE
Sync with dev branch

### PR DESCRIPTION
**Pull request by OptimumCode**
Add missing functionality to version 4.*

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -355,12 +371,6 @@ The `th2_check1_active_tasks_number` metric separate rules with label `rule_type
 + Migrated `sailfish-utils` version from `3.12.2` to `3.14.0`
     + sailfish updated to 3.3.54
 
-### 3.10.0
```
Comment (Nikita-Smirnov-Exactpro) [inline README.md:358]: Maybe save this records for history and add information about merging 3.10.0 for 4.4.0 version?

File: src/main/kotlin/com/exactpro/th2/check1/CollectorService.kt [hunk 2]
```diff
@@ -124,50 +138,125 @@ class CollectorService(
         ruleFactory = RuleFactory(configuration, streamObservable, eventBatchRouter)
     }
 
+    private fun prepareStoringResults(storeResult: Boolean): Triple<Long, CompletableFuture<Pair<EventStatus, Instant>>?, ((EventStatus) -> Unit)> {
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/check1/CollectorService.kt:141]: Please add TODO: replace triple and pair to specific classes

File: src/main/kotlin/com/exactpro/th2/check1/CollectorService.kt [hunk 3]
```diff
@@ -124,50 +138,125 @@ class CollectorService(
...
+                .apply { addToChainOrBegin(prevTask, request.checkpoint) }
                 .run { silenceCheckTask?.also { subscribeNextTask(it) } ?: this }
         }
-        return chainID
+
+        if (resultFuture != null) {
+            ruleIdToResult[ruleId] = if (silenceFuture == null)
+                resultFuture
+            else
+                resultFuture.thenCompose { if (it.first == EventStatus.SUCCESS) silenceFuture else resultFuture }
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/check1/CollectorService.kt:199]:
```suggestion
            ruleIdToResult[ruleId] = if (silenceFuture == null) {
                resultFuture
            } else {
                resultFuture.thenCompose { if (it.first == EventStatus.SUCCESS) silenceFuture else resultFuture }
            }
```

File: src/main/kotlin/com/exactpro/th2/check1/rule/AbstractCheckTask.kt [hunk 4]
```diff
@@ -85,7 +89,10 @@ abstract class AbstractCheckTask(
     private val eventBatchRouter: MessageRouter<EventBatch>
 ) : AbstractSessionObserver<MessageContainer>() {
 
-    protected open class Refs(val rootEvent: Event)
+    protected open class Refs(
+        val rootEvent: Event,
+        val onTaskFinished: ((EventStatus) -> Unit)
```
Comment (Nikita-Smirnov-Exactpro) [inline src/main/kotlin/com/exactpro/th2/check1/rule/AbstractCheckTask.kt:94]:
```suggestion
        val onTaskFinished: (EventStatus) -> Unit
```

**Review by Nikita-Smirnov-Exactpro — APPROVED** [review]

**Comment by OptimumCode** [comment]
All requested changes will be made in a separate PR

**Pull request merged**